Mathlib/MeasureTheory/Function/Holder.lean

```diff
@@ -0,0 +1,255 @@
+-/
+
+section SMul
+
+variable {α 𝕜' 𝕜 E : Type*} {m : MeasurableSpace α} {μ : Measure α}
+    {p q r : ℝ≥0∞} [hpqr : HolderTriple p q r]
+
+section MulActionWithZero
+
+variable [NormedRing 𝕜] [NormedAddCommGroup E] [MulActionWithZero 𝕜 E] [BoundedSMul 𝕜 E]
+
+/-- Heterogeneous scalar multiplication of `MeasureTheory.Lp` functions by `MeasureTheory.Lp`
+functions when the exponents satisfy `ENNReal.HolderTriple p q r`. -/
+instance : HSMul (Lp 𝕜 p μ) (Lp E q μ) (Lp E r μ) where
+  hSMul f g := (Lp.memℒp g).smul (Lp.memℒp f) |>.toLp (⇑f • ⇑g)
+
+lemma smul_def {f : Lp 𝕜 p μ} {g : Lp E q μ} :
+    f • g = ((Lp.memℒp g).smul (Lp.memℒp f)).toLp (⇑f • ⇑g) :=
+  rfl
+
+lemma coeFn_lp_smul (f : Lp 𝕜 p μ) (g : Lp E q μ) :
+    (f • g : Lp E r μ) =ᵐ[μ] ⇑f • g := by
+  rw [smul_def]
+  exact Memℒp.coeFn_toLp _
+
+protected lemma norm_smul_le (f : Lp 𝕜 p μ) (g : Lp E q μ) :
+    ‖f • g‖ ≤ ‖f‖ * ‖g‖ := by
+  simp only [Lp.norm_def, ← ENNReal.toReal_mul, coeFn_lp_smul]
+  refine ENNReal.toReal_mono ?_ ?_
+  · exact ENNReal.mul_ne_top (eLpNorm_ne_top f) (eLpNorm_ne_top g)
+  · rw [eLpNorm_congr_ae (coeFn_lp_smul f g)]
+    exact eLpNorm_smul_le_mul_eLpNorm (Lp.aestronglyMeasurable g) (Lp.aestronglyMeasurable f)
+
+end MulActionWithZero
+
+section Module
+
+variable [NormedRing 𝕜] [NormedAddCommGroup E] [Module 𝕜 E] [BoundedSMul 𝕜 E]
+
+protected lemma smul_add (f₁ f₂ : Lp 𝕜 p μ) (g : Lp E q μ) :
+    (f₁ + f₂) • g = f₁ • g + f₂ • g := by
+  simp only [smul_def, ← Memℒp.toLp_add]
+  apply Memℒp.toLp_congr
+  filter_upwards [AEEqFun.coeFn_add f₁.val f₂.val] with x hx
+  simp [hx, add_smul]
+
+protected lemma add_smul (f : Lp 𝕜 p μ) (g₁ g₂  : Lp E q μ) :
+    f • (g₁ + g₂) = f • g₁ + f • g₂ := by
+  simp only [smul_def, ← Memℒp.toLp_add]
+  apply Memℒp.toLp_congr _ _ ?_
+  filter_upwards [AEEqFun.coeFn_add g₁.val g₂.val] with x hx
+  simp [hx, smul_add]
+
+variable (E q) in
+@[simp]
+protected lemma smul_zero (f : Lp 𝕜 p μ) :
+    f • (0 : Lp E q μ) = (0 : Lp E r μ) := by
+  convert Memℒp.zero.toLp_zero
+  apply Memℒp.toLp_congr _ _ ?_
+  filter_upwards [Lp.coeFn_zero E q μ] with x hx
+  rw [Pi.smul_apply', hx]
+  simp
+
+variable (𝕜 p) in
+@[simp]
+protected lemma zero_smul (f : Lp E q μ) :
+    (0 : Lp 𝕜 p μ) • f = (0 : Lp E r μ) := by
+  convert Memℒp.zero.toLp_zero
+  apply Memℒp.toLp_congr _ _ ?_
+  filter_upwards [Lp.coeFn_zero 𝕜 p μ] with x hx
+  rw [Pi.smul_apply', hx]
+  simp
+
+@[simp]
+protected lemma smul_neg (f : Lp 𝕜 p μ) (g : Lp E q μ) :
+    f • -g = -(f • g) := by
+  simp [eq_neg_iff_add_eq_zero, ← Lp.add_smul]
+
+@[simp]
+protected lemma neg_smul (f : Lp 𝕜 p μ) (g : Lp E q μ) :
+    -f • g = -(f • g) := by
+  simp [eq_neg_iff_add_eq_zero, ← Lp.smul_add]
+
+protected lemma neg_smul_neg (f : Lp 𝕜 p μ) (g : Lp E q μ) :
+    -f • -g = f • g := by
+  simp
+
+variable [NormedRing 𝕜'] [Module 𝕜' E] [Module 𝕜' 𝕜] [BoundedSMul 𝕜' E] [BoundedSMul 𝕜' 𝕜]
+
+protected lemma smul_smul_assoc [IsScalarTower 𝕜' 𝕜 E]
```

Comment:
```suggestion
protected lemma smul_assoc [IsScalarTower 𝕜' 𝕜 E]
```

Reply:
Doesn't match `_root_.smul_assoc` (or any other `assoc` lemma).

Reply:
Sorry, but I disagree. It's the same shape as `_root_.smul_assoc`

Reply:
oops! I read the diff backwards.